File: Code/AutoDeployrBackend/analyzers/flask-ast-analyzer/analyzer.py
#!/usr/bin/env python3
import json
import sys
import os
import argparse
import logging
from ast_parser.flask_parser import FlaskApplicationParser
from ast_parser.http_method_fixer import fix_http_methods, fix_directory

#logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

def analyze_app(app_path, output_file=None, fix_methods=True):
    """Analyze Flask application and return or save result as JSON"""
    try:
        # Analysis
        parser = FlaskApplicationParser(app_path)
        result = parser.parse()

        # Convert to JSON
        result_json = json.dumps(result, indent=2)

        # Save or return
        if output_file:
            with open(output_file, 'w') as f:
                f.write(result_json)
            logger.info(f"Analysis saved to {output_file}")

            if fix_methods and os.path.exists(output_file):
                fix_http_methods(output_file)

            return 0
        else:
            print(result_json)
            return 0
    except Exception as e:
        logger.error(f"Error analyzing application: {str(e)}")
        print(json.dumps({"error": str(e)}))
        return 1

def analyze_file(file_path, output_file=None, fix_methods=True):
    """Analyze a single Python Flask file"""
    try:
        filename = os.path.basename(file_path)
        parser = FlaskApplicationParser("")
        result = parser.parse_file(file_path, filename)

        # Convert to JSON
        result_json = json.dumps(result, indent=2)

        # Save or return
        if output_file:
            with open(output_file, 'w') as f:
                f.write(result_json)
            logger.info(f"Analysis saved to {output_file}")

            # Fix HTTP methods if requested
            if fix_methods and os.path.exists(output_file):
                fix_http_methods(output_file)

            return 0
        else:
            print(result_json)
            return 0
    except Exception as e:
        logger.error(f"Error analyzing file: {str(e)}")
        print(json.dumps({"error": str(e)}))
        return 1

def fix_methods_in_directory(directory_path):
    """Fix HTTP methods in all Python files in a directory"""
    try:
        fixed_files = fix_directory(directory_path)
        if fixed_files:
            logger.info(f"Fixed HTTP methods in {len(fixed_files)} files:")
            for file in fixed_files:
                logger.info(f"  - {file}")
        else:
            logger.info("No files needed fixing.")
        return 0
    except Exception as e:
        logger.error(f"Error fixing HTTP methods: {str(e)}")
        return 1

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Analyze Flask applications using AST")
    parser.add_argument("--app-path", help="Path to Flask application directory")
    parser.add_argument("--file", help="Analyze a single Python file")
    parser.add_argument("--output", help="Output file for results (stdout if not specified)")
    parser.add_argument("--fix-methods", action="store_true", help="Fix HTTP methods in generated code")
    parser.add_argument("--fix-directory", help="Fix HTTP methods in all Python files in a directory")
    parser.add_argument("--no-fix", action="store_true", help="Don't fix HTTP methods in output files")

    args = parser.parse_args()

    # Default to fixing HTTP methods unless --no-fix is specified
    fix_methods = not args.no_fix

    if args.fix_directory:
        sys.exit(fix_methods_in_directory(args.fix_directory))
    elif args.app_path:
        sys.exit(analyze_app(args.app_path, args.output, fix_methods))
    elif args.file:
        sys.exit(analyze_file(args.file, args.output, fix_methods))
    else:
        parser.print_help()
        sys.exit(1)

File: Code/AutoDeployrBackend/analyzers/laravel-analyzer/analyzer.php
#!/usr/bin/env php
<?php

require_once __DIR__ . '/vendor/autoload.php';

use AutoDeployr\LaravelAnalyzer\LaravelApplicationParser;
use Symfony\Component\Console\Application;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;

class AnalyzeCommand extends Command
{
    protected static $defaultName = 'analyze';

    protected function configure()
    {
        $this
            ->setDescription('Analyze Laravel applications for serverless deployment')
            ->addOption('app-path', null, InputOption::VALUE_REQUIRED, 'Path to Laravel application directory')
            ->addOption('file', null, InputOption::VALUE_REQUIRED, 'Analyze a single PHP file')
            ->addOption('output', null, InputOption::VALUE_REQUIRED, 'Output file for results (stdout if not specified)')
            ->addOption('no-fix', null, InputOption::VALUE_NONE, 'Don\'t fix HTTP methods in output files');
    }

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        try {
            $appPath = $input->getOption('app-path');
            $file = $input->getOption('file');
            $outputFile = $input->getOption('output');
            $noFix = $input->getOption('no-fix');

            if ($appPath) {
                return $this->analyzeApp($appPath, $outputFile, !$noFix, $output);
            } elseif ($file) {
                return $this->analyzeFile($file, $outputFile, !$noFix, $output);
            } else {
                $output->writeln('<error>Either --app-path or --file must be specified</error>');
                return Command::FAILURE;
            }
        } catch (Exception $e) {
            $output->writeln('<error>Error: ' . $e->getMessage() . '</error>');
            return Command::FAILURE;
        }
    }

    private function analyzeApp(string $appPath, ?string $outputFile, bool $fixMethods, OutputInterface $output): int
    {
        try {
            $parser = new LaravelApplicationParser($appPath);
            $result = $parser->parse();

            $resultJson = json_encode($result, JSON_PRETTY_PRINT);

            if ($outputFile) {
                file_put_contents($outputFile, $resultJson);
                $output->writeln("Analysis saved to $outputFile");
            } else {
                echo $resultJson;
            }

            return Command::SUCCESS;
        } catch (Exception $e) {
            $output->writeln('<error>Error analyzing application: ' . $e->getMessage() . '</error>');
            echo json_encode(['error' => $e->getMessage()]);
            return Command::FAILURE;
        }
    }

    private function analyzeFile(string $filePath, ?string $outputFile, bool $fixMethods, OutputInterface $output): int
    {
        try {
            $filename = basename($filePath);
            $parser = new LaravelApplicationParser('');
            $result = $parser->parseFile($filePath, $filename);

            $resultJson = json_encode($result, JSON_PRETTY_PRINT);

            if ($outputFile) {
                file_put_contents($outputFile, $resultJson);
                $output->writeln("Analysis saved to $outputFile");
            } else {
                echo $resultJson;
            }

            return Command::SUCCESS;
        } catch (Exception $e) {
            $output->writeln('<error>Error analyzing file: ' . $e->getMessage() . '</error>');
            echo json_encode(['error' => $e->getMessage()]);
            return Command::FAILURE;
        }
    }
}

if (php_sapi_name() === 'cli' && isset($argv)) {
    $options = getopt('', ['app-path:', 'file:', 'output:', 'no-fix']);
    
    if (!empty($options)) {
        try {
            $appPath = $options['app-path'] ?? null;
            $file = $options['file'] ?? null;
            $outputFile = $options['output'] ?? null;
            $noFix = isset($options['no-fix']);

            if ($appPath) {
                $parser = new LaravelApplicationParser($appPath);
                $result = $parser->parse();
                $resultJson = json_encode($result, JSON_PRETTY_PRINT);
                
                if ($outputFile) {
                    file_put_contents($outputFile, $resultJson);
                    error_log("Analysis saved to $outputFile");
                } else {
                    echo $resultJson;
                }
                exit(0);
            } elseif ($file) {
                $parser = new LaravelApplicationParser('');
                $result = $parser->parseFile($file, basename($file));
                $resultJson = json_encode($result, JSON_PRETTY_PRINT);
                
                if ($outputFile) {
                    file_put_contents($outputFile, $resultJson);
                    error_log("Analysis saved to $outputFile");
                } else {
                    echo $resultJson;
                }
                exit(0);
            }
        } catch (Exception $e) {
            error_log("Error: " . $e->getMessage());
            echo json_encode(['error' => $e->getMessage()]);
            exit(1);
        }
    }
    $application = new Application('Laravel Analyzer', '1.0.0');
    $application->add(new AnalyzeCommand());
    $application->run();
} 